bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/parser/visitors/IdentifyAggAfterJoinVisitor.java

```diff
@@ -0,0 +1,143 @@
+import com.google.zetasql.parser.ASTNodes.ASTOnClause;
+import com.google.zetasql.parser.ASTNodes.ASTOnOrUsingClauseList;
+import com.google.zetasql.parser.ASTNodes.ASTPathExpression;
+import com.google.zetasql.parser.ASTNodes.ASTSelect;
+import com.google.zetasql.parser.ASTNodes.ASTTablePathExpression;
+import com.google.zetasql.parser.ParseTreeVisitor;
+import com.google.zetasql.toolkit.antipattern.util.ZetaSQLStringParsingHelper;
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+
+public class IdentifyAggAfterJoinVisitor extends ParseTreeVisitor {
+
+  private final String SUGGESTION_MESSAGE =
+      "GROUP By found at line number %s after Join at line number %s.Try to apply aggregation before joining to reduce join data";
+
+  private String query;
+  private boolean foundGroupBy = false;
+  private boolean foundJoin = false;
+
+  private int joinLocation=0;
+  private int groupByLocation=0;
+  private ArrayList<String> result = new ArrayList<String>();
+
+  private Set<String> lhsJoinKeys = new HashSet<>();
+  private Set<String> rhsJoinKeys = new HashSet<>();
+
+  private Set<String> groupByKeys = new HashSet<>();
+
+  public IdentifyAggAfterJoinVisitor(String query) {
+    this.query = query;
+  }
+
+  @Override
+  public void visit(ASTNodes.ASTSelect selectNode) {
+
+    // Traverse from the select node to check if a join exists and set the foundJoin flag
+    ASTNodes.ASTTableExpression tableExpression = selectNode.getFromClause().getTableExpression();
+    if (tableExpression instanceof ASTNodes.ASTJoin) {
+      foundJoin = true;
+      joinLocation = ZetaSQLStringParsingHelper.countLine(query, tableExpression.getParseLocationRange().start());
+
+      ASTOnClause onClause = ((ASTJoin) tableExpression).getOnClause();
+
+      if (onClause.getExpression() instanceof ASTNodes.ASTAndExpr) {
+        ASTNodes.ASTAndExpr andNode = (ASTNodes.ASTAndExpr) onClause.getExpression();
+        for (ASTNodes.ASTExpression elementNode : andNode.getConjuncts()) {
+          visit((ASTNodes.ASTBinaryExpression) elementNode);
+        }
+      } else {
+        ASTNodes.ASTBinaryExpression node = (ASTNodes.ASTBinaryExpression) onClause.getExpression();
+        visit(node);
+      }
+    }
+
+    // Check if group by exists and set the flag accordingly
+    if (foundJoin && selectNode.getGroupBy() != null) {
+
+      foundGroupBy = true;
+      groupByLocation = ZetaSQLStringParsingHelper.countLine(query, selectNode.getGroupBy().getParseLocationRange().start());
+      for (ASTGroupingItem groupingItem : selectNode.getGroupBy().getGroupingItems()) {
+        ASTNodes.ASTPathExpression groupPath = (ASTNodes.ASTPathExpression) groupingItem.getExpression();
+
+        if (groupPath.getNames().size() > 1) {
+          groupByKeys.add(groupPath.getNames().get(1).getIdString());
+        } else {
+          groupByKeys.add(groupPath.getNames().get(0).getIdString());
```

Comment:
What does the first entry of groupPath.getNames() correspond to, why do we need this value?

Reply:
When GROUP BY uses aliases, the first entry corresponds to the alias, else it corresponds to the column name. This is to ensure we are getting only the column name.

Reply:
it seems to me that you want the last record of the list. If so consider a syntax like the following 

```
ImmutableList<ASTIdentifier> namesList = lhsPath.getNames();
lhsJoinKeys.add(namesList.get(namesList.size()-1).getIdString());
```

Reply:
Added and tested the change suggested above.